pkg/deploy/controller/deploymentcleanup/controller.go

```diff
@@ -0,0 +1,181 @@
+	deployapi "github.com/openshift/origin/pkg/deploy/api"
+)
+
+const (
+	// deploymentConfigNameIndex is used to index replication controllers on the name of the deployment
+	// configuration that was used to create them
+	deploymentConfigNameIndex = "dc-name"
+)
+
+type DeploymentCleanupController struct {
+	deploymentConfigClient      osclient.DeploymentConfigsNamespacer
+	replicationControllerClient kclient.ReplicationControllersNamespacer
+
+	// deploymentConfigQueue holds the deployment configuration events that need to be checked
+	deploymentConfigQueue workqueue.RateLimitingInterface
+	maxRetries            int
+	// retries keeps track of how many times we've tried to respond to a deployment configuration
+	// event. Retries are keyed on namespace/name as in cache.MetaNamespaceKeyFunc
+	retries *lru.Cache
+
+	deploymentConfigCache      cache.Store
+	deploymentConfigController *framework.Controller
+
+	replicationControllerCache cache.Indexer
+}
+
+func (c *DeploymentCleanupController) Run(numWorkers int, stopChannel <-chan struct{}) {
+	defer kruntimeutil.HandleCrash()
+
+	// start controller to fill stores, call informers and populate queues
+	go c.deploymentConfigController.Run(stopChannel)
+
+	// poll stores until they've filled
+	for !c.deploymentConfigController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	// spawn worker threads to handle syncing
+	for i := 0; i < numWorkers; i++ {
+		go wait.Until(c.syncDeploymentConfigs, 0, stopChannel)
+	}
+
+	// block until the stop channel is closed
+	<-stopChannel
+
+	// shut down queues
+	c.deploymentConfigQueue.ShutDown()
+
+	// clear caches
+	c.retries.Purge()
+}
+
+// syncDeploymentConfigs triggers on deployment configuration update and add events in order
+// to delete old replication controller deployments if their quota has been reached
+func (c *DeploymentCleanupController) syncDeploymentConfigs() {
+	uncastKey, quit := c.deploymentConfigQueue.Get()
+	if quit {
+		return
+	}
+	// keys are string-valued `namespace/name,` generated with cache.MetaNamespaceKeyFunc
+	key := uncastKey.(string)
+
+	retry := false
+	// we always want to signal that we're done with this key after we process it
+	// but we may or may not want to re-try processing it later
+	defer func() {
+		if retry {
+			attempts, recorded := c.retries.Get(key)
+			if !recorded {
+				attempts = 0
+			}
+			c.retries.Add(key, attempts.(int)+1)
+			c.deploymentConfigQueue.AddRateLimited(key)
+		} else {
+			c.retries.Remove(key)
+		}
+
+		c.deploymentConfigQueue.Done(key)
+	}()
+
+	obj, exists, err := c.deploymentConfigCache.GetByKey(key)
+	if err != nil {
+		glog.V(2).Infof("Failed retrieving Deployment Config from cache with key %q: %v", key, err)
+		retry = c.canAffordRetrying(key)
+		return
+	}
+
+	if !exists {
+		// TODO: determine what to do here - should we allow manual fetch using the client calls?
+	}
+
+	deploymentConfig := obj.(*deployapi.DeploymentConfig)
```

Comment:
If `deploymentConfig.Spec.RevisionHistoryLimit == niil` does everything beyond this point become unnecessary work?